File: 2021/01/202101.py
def part1(data):
    mx = data[0]
    c = 0
    for i in range(1,len(data)):
        v = data[i]
        if v > mx:
            c += 1
            mx = v
        else:
            mx = v
    return c

td = [ 199, 200, 208, 210, 200, 207, 240, 269, 260, 263 ]
assert part1(td) == 7

data = [int(s) for s in open("input.txt").read().splitlines()]
print("PART1:", part1(data))

def sums(data):
    for i in range(0,len(data)-2):
        yield data[i] + data[i+1] + data[i+2]

def part2(data):
    return part1(list(sums(data)))

assert part2(td) == 5

print("PART2:", part2(data))

File: 2021/01/202101.exs
parse = fn x ->
  String.trim(x)
  |> String.split("\n")
  |> Enum.map(&String.to_integer/1)
end

{:ok, input} = File.read("input.txt")
data = parse.(input)

# enum based solution to p1 (4 lines!)
r1 =
  data
  |> Enum.chunk_every(2, 1, :discard)
  |> Enum.count(fn [x, y] -> y > x end)

IO.puts("PART1: #{r1}")

# recursion based solution to p1
defmodule Main do
  def p1r(ls) do
    case ls do
      [] -> 0
      [_x] -> 0
      [x, y | tail] when y > x -> 1 + p1r([y | tail])
      [_x, y | tail] -> p1r([y | tail])
    end
  end
end

IO.puts("PART1 recursive: #{Main.p1r(data)}")

r2 =
  data
  |> Enum.chunk_every(3, 1, :discard)
  |> Enum.map(&Enum.sum/1)
  |> Enum.chunk_every(2, 1, :discard)
  |> Enum.count(fn [x, y] -> y > x end)

IO.puts("PART2: #{r2}")
